lib/spack/spack/parser.py

```diff
@@ -76,7 +76,9 @@
 IDENTIFIER = r"([a-zA-Z_0-9][a-zA-Z_0-9\-]*)"
 DOTTED_IDENTIFIER = rf"({IDENTIFIER}(\.{IDENTIFIER})+)"
 GIT_HASH = r"([A-Fa-f0-9]{40})"
-GIT_VERSION = rf"((git\.({DOTTED_IDENTIFIER}|{IDENTIFIER}))|({GIT_HASH}))"
+#: Git refs include branch names, and can contain "." and "/"
+GIT_REF = r"([a-zA-Z_0-9][a-zA-Z_0-9./\-]*)"
```

Comment:
```suggestion
GIT_REF = r"([a-zA-Z_0-9][a-zA-Z_0-9./-]*)"
```

In fact it's more strict, from `man git-check-ref-format`: no repeated `.` or `/`, cannot end with `.` or `/`. The current regex is not a subset of valid git refs. Something along the lines of

```
[a-zA-Z_0-9](?:[a-zA-Z_0-9-]|(?:\.(?!\.)|/(?!/))(?=[a-zA-Z_0-9-]))*
```

probably is

Reply:
I can get rid of the `\` but I don't want our spec parser to throw I-don't-recognize-this-and-can't-parse-it-errors because the user didn't enter an appropriate git ref: I want to leave that to the part that attempts to pull the ref (later, in the fetcher)

Reply:
Actually, when I first wrote this, I think it was to escape the `-`. While https://docs.python.org/3/library/re.html does not require this (if `-` is at the end of the character set), we already do it for `IDENTIFIER` (and `NAME`, although it's required there since `-` is not at the end of the set). IMO it therefore makes sense to keep this (I don't want to have two different ways of handling `-` in character sets, and I don't want to change those regexes in this PR.

Reply:
That's alright with me